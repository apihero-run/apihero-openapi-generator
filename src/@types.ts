export type Enum = {
  name: string;
  value: string;
  type: string;
  description: string | null;
};

export type Model = Schema & {
  name: string;
  export:
    | "reference"
    | "generic"
    | "enum"
    | "array"
    | "dictionary"
    | "interface"
    | "one-of"
    | "any-of"
    | "all-of";
  type: string;
  base: string;
  template: string | null;
  link: Model | null;
  description: string | null;
  deprecated?: boolean;
  default?: string;
  imports: string[];
  enum: Enum[];
  enums: Model[];
  properties: Model[];
};

export type ModelComposition = {
  type: "one-of" | "any-of" | "all-of";
  imports: string[];
  enums: Model[];
  properties: Model[];
};

export type Schema = {
  isDefinition: boolean;
  isReadOnly: boolean;
  isRequired: boolean;
  isNullable: boolean;
  format?: string;
  maximum?: number;
  exclusiveMaximum?: boolean;
  minimum?: number;
  exclusiveMinimum?: boolean;
  multipleOf?: number;
  maxLength?: number;
  minLength?: number;
  pattern?: string;
  maxItems?: number;
  minItems?: number;
  uniqueItems?: boolean;
  maxProperties?: number;
  minProperties?: number;
};

export type OperationParameter = Model & {
  in: "path" | "query" | "header" | "formData" | "body" | "cookie";
  prop: string;
  mediaType: string | null;
};

export type OperationParameters = {
  imports: string[];
  parameters: OperationParameter[];
  parametersPath: OperationParameter[];
  parametersQuery: OperationParameter[];
  parametersForm: OperationParameter[];
  parametersCookie: OperationParameter[];
  parametersHeader: OperationParameter[];
  parametersBody: OperationParameter | null;
};

export type OperationError = {
  code: number;
  description: string;
};

export type OperationResponse = Model & {
  in: "response" | "header";
  code: number;
};

export type Operation = OperationParameters & {
  service: string;
  name: string;
  summary: string | null;
  description: string | null;
  deprecated: boolean;
  method: string;
  path: string;
  errors: OperationError[];
  results: OperationResponse[];
  responseHeader: string | null;
};

export type Service = {
  name: string;
  operations: Operation[];
  imports: string[];
};

export type Client = {
  version: string;
  models: Model[];
  services: Service[];
};

export type Type = {
  type: string;
  base: string;
  template: string | null;
  imports: string[];
  isNullable: boolean;
};

export interface IFileWriter {
  writeFile(path: string, data: string): Promise<void>;
}
