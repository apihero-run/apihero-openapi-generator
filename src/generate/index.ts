import { camelCase } from "camel-case";
import { Client, Model, Operation, Service } from "../@types";

export type GenerateClientOptions = {
  additionalImports?: Array<{ imports: string[]; name: string }>;
  additionalData?: Record<string, string | number>;
};

export function generateFromClient(
  client: Client,
  options?: GenerateClientOptions,
): Map<string, string> {
  const files = new Map<string, string>();

  const modelCode = generateClientModels(client.models);

  files.set("@types.ts", modelCode);

  const servicesCode = generateClientServices(client.services, options);

  for (const [name, code] of servicesCode) {
    files.set(`${name}.ts`, code);
  }

  const indexCode = generateClientIndex(servicesCode);

  files.set("index.ts", indexCode);

  return files;
}

export function generateOperationCodeFromClient(client: Client, operationId: string): string {
  let operation: Operation | undefined;

  for (const service of client.services) {
    for (const op of service.operations) {
      if (op.id === operationId) {
        operation = op;
        break;
      }
    }
  }

  if (!operation) {
    throw new Error(`Operation with id ${operationId} not found`);
  }

  const models = getAllImportsRecursive(operation.imports, client.models);

  const modelCode = generateClientModels(models);

  const operationCode = generateOperation(operation);

  return `${modelCode}\n\n${operationCode}`;
}

function generateClientIndex(servicesCode: ReturnType<typeof generateClientServices>): string {
  const imports = servicesCode
    .map(([name]) => `import * as ${camelCase(name)} from "./${name}";`)
    .join("\n");

  const exports = `export { ${servicesCode.map(([name]) => camelCase(name)).join(", ")} };`;

  return `${imports}\n\n${exports}\n\nexport * from "./@types";`;
}

function generateClientServices(
  services: Client["services"],
  options?: GenerateClientOptions,
): Array<[string, string]> {
  return services.map((service) => generateClientService(service, options));
}

function generateClientService(
  service: Service,
  options?: GenerateClientOptions,
): [string, string] {
  const serviceCode = generateClientServiceCode(service, options);

  return [service.name, serviceCode];
}

function generateClientServiceCode(service: Service, options?: GenerateClientOptions): string {
  const imports = generateServiceImports(service, options?.additionalImports);

  return `${imports}\n\n${service.operations
    .map((operation) => generateOperation(operation, options?.additionalData))
    .join("\n\n")}`;
}

function generateServiceImports(
  service: Service,
  additionalImports?: GenerateClientOptions["additionalImports"],
): string {
  const additionalImportsCode = (additionalImports ?? [])
    .map((additionalImport) => {
      return `import { ${additionalImport.imports.join(", ")} } from "${additionalImport.name}";`;
    })
    .join("\n");

  const serviceImportsCode =
    service.imports.length > 0
      ? `import { ${service.imports.join(", ")}, ApiHeroEndpoint } from "./@types";`
      : "";

  return `${additionalImportsCode}\n${serviceImportsCode}`;
}

function generateOperation(
  operation: Operation,
  additionalData?: GenerateClientOptions["additionalData"],
): string {
  return `${generateOperationDocs(operation)}\n${generateOperationExport(
    operation,
    additionalData,
  )}`;
}

function generateOperationExport(
  operation: Operation,
  additionalData?: GenerateClientOptions["additionalData"],
): string {
  const parameters = generateOperationParameters(operation);
  const result = generateOperationResult(operation);
  const headers = generateOperationHeaders(operation);

  return `export const ${operation.name}: ApiHeroEndpoint<${parameters}, ${result}${
    headers ? `, ${headers}` : ``
  }> = {
    id: '${operation.id}',
    ${
      additionalData
        ? `${Object.entries(additionalData)
            .map(([key, value]) => `${key}: ${JSON.stringify(value)}`)
            .join(",\n")},\n`
        : ""
    }
  }`;
}

function generateOperationResult(operation: Operation): string {
  if (!operation.results) {
    return "void";
  }

  return operation.results.map((result) => generateType(result)).join(" | ");
}

function generateOperationHeaders(operation: Operation): string | undefined {
  const allHeaders = operation.results.flatMap((result) => result.headers);

  if (allHeaders.length === 0) {
    return;
  }

  return operation.results
    .flatMap((result) => {
      return `{ ${result.headers
        .map((header) => `"${header.name}": ${generateType(header)}; `)
        .join(" ")} }`;
    })
    .join(" | ");
}

function generateOperationParameters(operation: Operation): string {
  if (!operation.parameters || operation.parameters.length === 0) {
    return "never";
  }

  return `{ ${operation.parameters.map(generateOperationParameter).join("; ")} }`;
}

function generateOperationParameter(parameter: Operation["parameters"][0]): string {
  return `${parameter.name}${parameter.isRequired ? "" : "?"}: ${generateType(parameter)}`;
}

function generateOperationParameterComments(parameters: Model[], path: string[] = []): string {
  if (!parameters || parameters.length === 0) {
    return "";
  }

  return `\n\n${parameters
    .map((parameter) => generateOperationParameterComment(parameter, path))
    .join("\n")}`;
}

function generateOperationParameterComment(parameter: Model, path: string[] = []): string {
  return `* @param ${[...path, parameter.name].join(".")} ${
    parameter.description ? `- ${parameter.description}` : ""
  }${generateOperationParameterComments(parameter.properties, [parameter.name])}`;
}

function generateOperationDocs(operation: Operation): string {
  const summeryAndDescriptionEqual = operation.summary === operation.description;

  if (summeryAndDescriptionEqual) {
    return `/** ${operation.externalDocs ? `\n* @see ${operation.externalDocs}\n` : "\n"}${
      operation.deprecated ? "\n* @deprecated\n" : "\n"
    }${
      operation.summary ? "\n* " + escapeComment(operation.summary) : ""
    }${generateOperationParameterComments(operation.parameters)} \n*/`;
  }

  return `/** ${operation.externalDocs ? `\n* @see ${operation.externalDocs}\n` : "\n"}${
    operation.deprecated ? "\n* @deprecated\n" : "\n"
  }${operation.summary ? "\n* " + escapeComment(operation.summary) : ""}${
    operation.description ? "\n* " + escapeComment(operation.description) : ""
  }${generateOperationParameterComments(operation.parameters)} \n*/`;
}

const escapeComment = (value: string): string => {
  return value
    .replace(/\*\//g, "*")
    .replace(/\/\*/g, "*")
    .replace(/\r?\n(.*)/g, (_, w) => `\n * ${w.trim()}`);
};

function generateClientModels(models: Client["models"]): string {
  const endpointGenericCode = `export type ApiHeroEndpoint<Params, ResponseBody, Headers = unknown> = \n{ id: string;\n[key: string]: string | number;\n };`;

  const modelsCode = models.map(generateClientModel).join("\n\n");

  return `${endpointGenericCode}\n\n${modelsCode}`;
}

export function generateClientModel(model: Model): string {
  if (model.export === "interface") {
    return generateClientModelInterface(model);
  }

  return generateClientModelGeneric(model);
}

function generateClientModelGeneric(model: Model): string {
  return `${generateModelComments(model)}\nexport type ${model.name} = ${generateType(model)}`;
}

function generateClientModelInterface(model: Model): string {
  return `${generateModelComments(model)}\nexport type ${model.name} = { ${generateModelProperties(
    model,
  )}}`;
}

function generateModelComments(m: Model) {
  if (!m.description && !m.deprecated && (!m.examples || m.examples.length === 0)) {
    return "";
  }

  return (
    "\n\n/** " +
    (m.description ? `\n* ${m.description}` : "") +
    (m.deprecated ? `\n* @deprecated` : "") +
    generateModelExamples(m) +
    "\n*/"
  );
}

function prettyPrintJSONInComment(json: string): string {
  return `${json.replace(/\r?\n/g, "\n* ")}`;
}

function generateModelExamples(m: Model) {
  if (!m.example && (!m.examples || m.examples.length === 0)) {
    return "";
  }

  if (m.example) {
    return `\n* @example\n* ${prettyPrintJSONInComment(JSON.stringify(m.example, null, 2))}`;
  }

  if (m.examples) {
    return `\n* ${m.examples
      .map(
        (example) =>
          `\n* @example\n* ${prettyPrintJSONInComment(JSON.stringify(example, null, 2))}`,
      )
      .join("\n* ")}`;
  }

  return "";
}

function generateType(property: Model, parent?: Model) {
  switch (property.export) {
    case "interface":
      return generateTypeInterface(property, parent);
    case "generic":
      return generateTypeGeneric(property, parent);
    case "enum":
      return generateTypeEnum(property, parent);
    case "array":
      return generateTypeArray(property, parent);
    case "reference":
      return generateTypeReference(property, parent);
    case "dictionary":
      return generateTypeDictionary(property, parent);
    case "one-of":
    case "any-of":
      return generateCompositeType(property, "|", parent);
    case "all-of":
      return generateCompositeType(property, "&", parent);
    default:
      throw new Error(`Unknown export type: ${property.export}`);
  }
}

function generateCompositeType(model: Model, compositeType: "|" | "&", parent?: Model): string {
  const unique = <T>(val: T, index: number, arr: T[]): boolean => {
    return arr.indexOf(val) === index;
  };

  const types = model.properties.map((property) => generateType(property, parent));
  const uniqueTypes = types.filter(unique);
  let uniqueTypesString = uniqueTypes.join(` ${compositeType} `);
  if (uniqueTypes.length > 1) {
    uniqueTypesString = `(${uniqueTypesString})`;
  }
  return uniqueTypesString;
}

function generateTypeDictionary(property: Model, parent?: Model): string {
  if (property.link) {
    return `Record<string, ${generateType(property.link)}>${generateIsNullable(property)}`;
  } else {
    return `Record<string, ${property.base}>${generateIsNullable(property)}`;
  }
}

function generateTypeReference(property: Model, parent?: Model): string {
  return `${property.base}${generateIsNullable(property)}`;
}

function generateTypeArray(property: Model, parent?: Model): string {
  if (property.link) {
    return `Array<${generateType(property.link)}>${generateIsNullable(property)}`;
  } else {
    return `Array<${property.base}>${generateIsNullable(property)}`;
  }
}

function generateTypeEnum(property: Model, parent?: Model): string {
  return `${property.enum.map((e) => e.value).join(" | ")}${generateIsNullable(property)}`;
}

function generateTypeGeneric(property: Model, parent?: Model): string {
  return `${generateGenericBase(property.base)}${generateIsNullable(property)}`;
}

function generateGenericBase(base: string): string {
  return base === "binary" ? "ReadableStream" : base;
}

function generateTypeInterface(property: Model, parent?: Model): string {
  if (!property.properties) {
    return "any";
  } else {
    return `{
${property.properties.map((p) => generateModelProperty(p, parent)).join("")}
}${generateIsNullable(property)}`;
  }
}

function generateModelProperty(property: Model, parent?: Model): string {
  return `${generateModelComments(property)}\n${property.isReadOnly ? "readonly " : ""}${
    property.name
  }${!property.isRequired && !property.default ? "?" : ""}: ${generateType(property, parent)};`;
}

function generateModelProperties(model: Model) {
  return model.properties.map((p) => generateModelProperty(p, model)).join("");
}

function generateIsNullable(property: Model): string {
  return property.isNullable ? " | null" : "";
}

function getAllImportsRecursive(
  imports: string[],
  allModels: Model[],
  models: Model[] = [],
): Model[] {
  const newModels = imports
    .map((i) => allModels.find((model) => model.name === i))
    .filter((m) => m !== undefined) as Model[];

  if (newModels.length === 0) {
    return models;
  }

  const newImports = newModels
    .flatMap((m) => m.imports)
    .filter(unique)
    .filter((i) => !imports.includes(i))
    .filter((i) => !models.find((m) => m.name === i))
    .filter((i) => !newModels.find((m) => m.name === i))
    .sort(sort);

  return getAllImportsRecursive(newImports, allModels, models.concat(newModels));
}

const unique = <T>(val: T, index: number, arr: T[]): boolean => {
  return arr.indexOf(val) === index;
};

const sort = (a: string, b: string): number => {
  const nameA = a.toLowerCase();
  const nameB = b.toLowerCase();
  return nameA.localeCompare(nameB, "en");
};
