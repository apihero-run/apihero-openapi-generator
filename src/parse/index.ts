import { OpenAPI } from "openapi-types";
import { Client } from "../@types";
import { getModels } from "./models";
import { getServices } from "./services";

export function parseSpec(doc: OpenAPI.Document): Client {
  return {
    version: doc.info.version,
    models: getModels(doc),
    services: getServices(doc),
  };
}
