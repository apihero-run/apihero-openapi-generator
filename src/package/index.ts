import path from "node:path";
import { mkdir, readdir, stat, unlink, writeFile } from "node:fs/promises";
import * as ts from "typescript";
import { GenerationOptions } from "../generate";

export type GeneratePackageOptions = {
  name: string;
  version: { major: number; minor: number; patch: number };
  description?: string;
  generation?: GenerationOptions;
};

export async function generatePackageFromTypeScriptFiles(
  files: Map<string, string>,
  destination: string,
  options: GeneratePackageOptions,
) {
  // Create the destination directory if it doesn't exist
  await mkdir(destination, { recursive: true });

  // Remove existing files in the destination directory
  const filesInDestination = await readdir(destination);

  for (const file of filesInDestination) {
    const fileStat = await stat(path.join(destination, file));

    if (fileStat.isFile()) {
      await unlink(path.join(destination, file));
    }
  }

  const packageJson = {
    name: `@apihero/${options.name}`,
    version: `${options.version.major}.${options.version.minor}.${options.version.patch}`,
    description: options.description,
    main: "./lib/index.js",
    exports: "./lib/index.js",
    types: "./lib/index.d.ts",
    files: ["/lib"],
    engines: {
      node: ">=14.0.0",
    },
    author: "API Hero",
    license: "MIT",
    keywords: ["api", options.name, "apihero-client", "http"],
    publishConfig: {
      access: "public",
    },
    repository: {
      type: "git",
      url: "https://github.com/jsonhero-io/apihero",
    },
    homepage: `https://apihero.run/docs/integrations/${options.name}`,
    bugs: {
      url: "https://github.com/jsonhero-io/apihero/issues",
    },
  };

  const packageJsonPath = path.join(destination, "package.json");
  const packageJsonContent = JSON.stringify(packageJson, null, 2);

  await writeFile(packageJsonPath, packageJsonContent);

  const typescriptPath = path.join(destination, "src");

  // Make sure the typescript path exists
  await mkdir(typescriptPath, { recursive: true });

  const filePaths = [];

  for (const [file, content] of Array.from(files)) {
    const filePath = path.join(typescriptPath, file);
    console.log(`Writing ${filePath}`);
    await writeFile(filePath, content);

    filePaths.push(filePath);
  }

  const compileResult = compile(filePaths, {
    outDir: path.join(destination, "lib"),
    target: ts.ScriptTarget.ES5,
    module: ts.ModuleKind.CommonJS,
    declaration: true,
  });

  if (!compileResult) {
    throw new Error("Compilation failed");
  }

  console.log("Compilation succeeded");

  // Create the README file
  const readmePath = path.join(destination, "README.md");
  const readmeContent = `# ${options.name}`;

  await writeFile(readmePath, readmeContent);

  // Create the license file

  const licensePath = path.join(destination, "LICENSE");
  const licenseContent = `Copyright (c) ${new Date().getFullYear()} API Hero

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
  `;

  await writeFile(licensePath, licenseContent);
}

function compile(fileNames: string[], options: ts.CompilerOptions): boolean {
  const program = ts.createProgram(fileNames, options);
  const emitResult = program.emit();

  const allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);

  allDiagnostics.forEach((diagnostic) => {
    if (diagnostic.file) {
      const { line, character } = ts.getLineAndCharacterOfPosition(
        diagnostic.file,
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        diagnostic.start!,
      );
      const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
      console.log(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
    } else {
      console.log(ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n"));
    }
  });

  return !emitResult.emitSkipped;
}
