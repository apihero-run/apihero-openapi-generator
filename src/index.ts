import { OpenAPI } from "openapi-types";
import { Client, Enum, Model, Operation, Service } from "./@types";
import {
  generateFromClient,
  GenerateClientOptions,
  GenerateClientCodeResult,
  GenerationOptions,
} from "./generate";
import { parseSpec } from "./parse";
import { format } from "prettier";

export { GenerateClientOptions, GenerationOptions };

export function generateClientFiles(
  doc: OpenAPI.Document,
  identifier: string,
  options?: GenerateClientOptions,
) {
  const client = generateClient(doc);

  return generateClientCode(client, identifier, options);
}

export function generateClientCode(
  client: Client,
  identifier: string,
  options?: GenerateClientOptions,
): GenerateClientCodeResult {
  const { files, mappings } = generateFromClient(client, identifier, options);

  // Format each file
  for (const [name, code] of Object.entries(files)) {
    files[name] = format(code, { parser: "typescript", printWidth: 100 });
  }

  return { files, mappings };
}

export function generateClient(doc: OpenAPI.Document): Client {
  return postProcessClient(parseSpec(doc));
}

const postProcessClient = (client: Client): Client => {
  return {
    ...client,
    models: client.models.map((model) => postProcessModel(model)),
    services: client.services.map((service) => postProcessService(service)),
  };
};

const postProcessService = (service: Service): Service => {
  const clone = { ...service };
  clone.operations = postProcessServiceOperations(clone);
  clone.operations.forEach((operation) => {
    clone.imports.push(...operation.imports);
  });
  clone.imports = postProcessServiceImports(clone);
  return clone;
};

export const postProcessServiceImports = (service: Service): string[] => {
  return service.imports.filter(unique).sort(sort);
};

const unique = <T>(val: T, index: number, arr: T[]): boolean => {
  return arr.indexOf(val) === index;
};

const sort = (a: string, b: string): number => {
  const nameA = a.toLowerCase();
  const nameB = b.toLowerCase();
  return nameA.localeCompare(nameB, "en");
};

export const postProcessServiceOperations = (service: Service): Operation[] => {
  const names = new Map<string, number>();

  return service.operations.map((operation) => {
    const clone = JSON.parse(JSON.stringify(operation)) as Operation;

    // Parse the service parameters and results, very similar to how we parse
    // properties of models. These methods will extend the type if needed.
    clone.imports.push(...flatMap(clone.parameters, (parameter) => parameter.imports));
    clone.imports.push(...flatMap(clone.results, (result) => result.imports));

    // Make sure the imports are unique
    clone.imports = clone.imports.filter(unique).sort(sort);

    // Check if the operation name is unique, if not then prefix this with a number
    const name = clone.name;
    const index = names.get(name) || 0;
    if (index > 0) {
      clone.name = `${name}${index}`;
    }
    names.set(name, index + 1);

    return clone;
  });
};

const flatMap = <U, T>(array: T[], callback: (value: T, index: number, array: T[]) => U[]): U[] => {
  const result: U[] = [];
  array.map<U[]>(callback).forEach((arr) => {
    result.push(...arr);
  });
  return result;
};

const postProcessModel = (model: Model): Model => {
  const postProcessModelImports = (model: Model): string[] => {
    return model.imports
      .filter(unique)
      .sort(sort)
      .filter((name) => model.name !== name);
  };

  const postProcessModelEnums = (model: Model): Model[] => {
    return model.enums.filter((property, index, arr) => {
      return arr.findIndex((item) => item.name === property.name) === index;
    });
  };

  const postProcessModelEnum = (model: Model): Enum[] => {
    return model.enum.filter((property, index, arr) => {
      return arr.findIndex((item) => item.name === property.name) === index;
    });
  };

  const sort = (a: string, b: string): number => {
    const nameA = a.toLowerCase();
    const nameB = b.toLowerCase();
    return nameA.localeCompare(nameB, "en");
  };

  const unique = <T>(val: T, index: number, arr: T[]): boolean => {
    return arr.indexOf(val) === index;
  };

  return {
    ...model,
    imports: postProcessModelImports(model),
    enums: postProcessModelEnums(model),
    enum: postProcessModelEnum(model),
  };
};
