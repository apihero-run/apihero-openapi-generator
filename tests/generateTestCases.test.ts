import { OpenAPI } from "openapi-types";
import { generateClientFiles } from "../src";
import { readFile } from "fs/promises";

async function loadTestCase(testCaseId: string): Promise<OpenAPI.Document> {
  const rawFile = await readFile(`./tests/fixtures/testCases/${testCaseId}.json`, "utf8");

  return JSON.parse(rawFile) as OpenAPI.Document;
}

async function generateTestCase(testCaseId: string): Promise<Map<string, string>> {
  const doc = await loadTestCase(testCaseId);

  const code = generateClientFiles(doc, testCaseId);

  return code;
}

test("code generation test cases", async () => {
  expect(await generateTestCase("createLabel")).toMatchSnapshot();
});
