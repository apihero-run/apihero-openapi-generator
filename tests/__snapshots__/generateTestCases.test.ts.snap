// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generation test cases 1`] = `
Map {
  "@types.ts" => "export type ApiHeroEndpoint<Params, ResponseBody, Headers = unknown> = {
  id: string;
  [key: string]: string | number;
};

/**
 * Basic Error
 */
export type BasicError = {
  url?: string;
  status?: string;
  message?: string;
  documentation_url?: string;
};

/**
 * Validation Error
 */
export type ValidationError = {
  errors?: Array<{
    code: string;
    field?: string;
    index?: number;
    value?: string | null | number | null | Array<string> | null;
    message?: string;
    resource?: string;
  }>;
  message: string;
  documentation_url: string;
};

/**
 * Color-coded labels help you categorize and filter your issues (just like labels in Gmail).
 * @example @see https://apihero.run/integrations/createLabel/examples/label
 */
export type Label = {
  /**
   *
   * @example
   * 208045946
   */
  id: number;

  /**
   * URL for the label
   *
   * @example
   * \\"https://api.github.com/repositories/42/labels/bug\\"
   */
  url: string;

  /**
   * The name of the label.
   *
   * @example
   * \\"bug\\"
   */
  name: string;

  /**
   * 6-character hex code, without the leading #, identifying the color
   *
   * @example
   * \\"FFFFFF\\"
   */
  color: string;

  /**
   *
   * @example
   * true
   */
  default: boolean;

  /**
   *
   * @example
   * \\"MDU6TGFiZWwyMDgwNDU5NDY=\\"
   */
  node_id: string;

  /**
   *
   * @example
   * \\"Something isn't working\\"
   */
  description: string;
};
",
  "issues.ts" => "import { Label, ApiHeroEndpoint } from \\"./@types\\";

/** 
* @see https://docs.github.com/rest/reference/issues#create-a-label


* Create a label

* @param repo - The name of the repository. The name is not case sensitive.
* @param owner - The account owner of the repository. The name is not case sensitive.
* @param body 

* @param body.name - The name of the label. Emoji can be added to label names, using either native emoji or colon-style markup. For example, typing \`:strawberry:\` will render the emoji ![:strawberry:](https://github.githubassets.com/images/icons/emoji/unicode/1f353.png \\":strawberry:\\"). For a full list of available emoji and codes, see \\"[Emoji cheat sheet](https://github.com/ikatyang/emoji-cheat-sheet).\\"
* @param body.color - The [hexadecimal color code](http://www.color-hex.com/) for the label, without the leading \`#\`.
* @param body.description - A short description of the label. Must be 100 characters or fewer. 
*/
export const createLabel: ApiHeroEndpoint<
  {
    repo: string;
    owner: string;
    body: {
      /**
       * The name of the label. Emoji can be added to label names, using either native emoji or colon-style markup. For example, typing \`:strawberry:\` will render the emoji ![:strawberry:](https://github.githubassets.com/images/icons/emoji/unicode/1f353.png \\":strawberry:\\"). For a full list of available emoji and codes, see \\"[Emoji cheat sheet](https://github.com/ikatyang/emoji-cheat-sheet).\\"
       */
      name: string;

      /**
       * The [hexadecimal color code](http://www.color-hex.com/) for the label, without the leading \`#\`.
       */
      color?: string;

      /**
       * A short description of the label. Must be 100 characters or fewer.
       */
      description?: string;
    };
  },
  Label,
  { Location: string }
> = {
  id: \\"issues/create-label\\",
};
",
  "index.ts" => "import * as issues from \\"./issues\\";

export { issues };

export * from \\"./@types\\";
",
}
`;
